### ISSUE TITLE
[1.19.2] [Forge] Compatibility mode causes land biomes to generate with ocean-like terrain

### ISSUE THREAD

**Issue by SandwichHorror** [Cristelknight/Expanded-Ecosphere#8]
**Versions Used:**
- Minecraft 1.19.2
- Forge 43.1.47
- wwoo_forge-3.1.8-1.19.2
- TerraBlender-forge-1.19.2-2.0.1.130

### **Issue:**
Enabling **"compatibility mode"** in the WWOO config causes land biomes to generate with ocean-like terrain.
This can be replicated with only the mods above. Simply teleport/fly around and eventually you'll find land biomes that generate as flat masses of water.

Here you can see multiple biomes that have all generated as "oceans".

![compat-biomes](https://user-images.githubusercontent.com/59847281/202878542-fd7441cb-64f6-4d62-b971-fd492d2fcaa2.PNG)

If this can't easily be fixed, then a way to disable specific WWOO biomes without using compatibility mode would be greatly appreciated.

**Comment by XAURAVUS** [comment]
I have also come across this issue - here's some screenshots from my instance demonstrating the same bug:
![2022-11-11_18 30 41](https://user-images.githubusercontent.com/82698242/202882173-5aec5490-aa67-4d80-8472-277180188ae9.png)
![2022-11-11_18 30 53](https://user-images.githubusercontent.com/82698242/202882254-23b9f7ff-d9e6-4159-8a85-196bddfb0f17.png)
![2022-11-11_18 31 23](https://user-images.githubusercontent.com/82698242/202882318-f70976ac-1076-4304-aad3-2916f7d685c4.png)

**Comment by Cristelknight** [comment]
It is a known problem. For some reason terrablender puts biomes which aren’t ocean biomes into oceans. William said, he has a idea what might cause it. So this issue will be fixed in the future I guess.

**Comment by SandwichHorror** [comment]
The interesting thing is it doesn't happen with Terrablender and WWOO **default** mode, only **compatible** mode.

Since Terrablender alone isn't the issue, it means the absence or presence of some setting in the compatible mode datapack is responsible for the weird ocean biomes.

Taking a quick look between the compatible and default datapacks, it looks like the default datapack has robust noise settings for the overworld, while compatible mode doesn't. Could this be part of the issue?

**Comment by Cristelknight** [comment]
No. That’s not the problem. In default mode biomes are added trough the overworld.json file. In compatible they are added with terrablender.

**Comment by zachsai** [comment]
I have this happening too when BOP is installed with William withers
i have it in default mode currently.
Deep Frozen Ocean spawns with land biomes

**Comment by TheButterbrotMan** [comment]
I am experiencing this problem also, mainly with structures. Is there any news on this?

**Comment by Mecheon** [comment]
Can confirm experienced this issue on 1.20.1 with most recent version of mods involved

**Comment by nickcorrado** [comment]
@Cristelknight Hello, I have a fix for this issue. The cause is that the Terrablender compatibility code works by tossing out all the vanilla biomes from the biome table before handing it over. But those vanilla biomes occupied important ranges in the overall biome space, most obviously the ocean biomes. The game selects biomes by taking the nearest available point in the space, so by removing the vanilla ocean biomes, in Terrablender's Wythers regions, the best it can do is to grab a Wythers *land* biome. Anyway, this is the offending line:

https://github.com/Cristelknight/Expanded-Ecosphere/blob/architectury-1.21.1/common/src/main/java/de/cristelknight/wwee/terra/TerraInit.java#L54

Remove that one line and biomes are placed perfectly.

I have no idea why the vanilla biomes were filtered out when creating compatibility with Terrablender; possibly it was because by default Terrablender adds a vanilla biome region, too, and so you'd end up with twice as many vanilla biomes. But this is easily fixed by just telling Terrablender not to do that in config. Instead of breaking Wythers biome placement on compatibility mod, just tell people to change `vanilla_overworld_region_weight = 10` to `0` in the Terrablender config.

I'd be happy to open a PR for this, but I dunno whether you're still maintaining the mod.

**Comment by Cristelknight** [comment]
Are you sure that’s how Terrablender works? I mean that sounds like nonsense to me. That would mean that this would also happen to mods like Biomes O Plenty and that they have to add vanilla biomes to their regions too? Like what?

Why would it remove all the vanilla biomes from a region I thought it would just override the vanilla biomes in the region where the mod actually provides a custom, new biome.

**Comment by nickcorrado** [comment]
BOP *does* add the vanilla biomes to its regions; see e.g. https://github.com/Glitchfiend/BiomesOPlenty/blob/26.2/common/src/main/java/biomesoplenty/biome/BOPOverworldBiomeBuilder.java where it creates tables of vanilla biomes and BOP biomes for various positions within the climate space. Its `add...` methods below sometimes include vanilla biomes, e.g. `addOffCoastBiomes` includes the BOP island biomes, vanilla's mushroom fields, and vanilla's ocean biomes. And even most of the other methods are designed to try to add a BOP biome first, then fall back to a vanilla biome if none is available; that's doing most of the "overriding vanilla biomes" work, but clearly doesn't apply to every case.

Same with others like Nature's Spirit. It defines a bunch of regions; here's one I picked at random: https://github.com/Team-Hibiscus/NaturesSpirit/blob/1.20.1/src/main/java/net/hibiscus/naturespirit/terrablender/TerraSolarisParameters.java It's accomplishing basically the same thing as BOP just with more config. Notice on line 131 how ocean biomes are literally just the vanilla ones. Or how commonBiomes includes a mixture of vanilla biomes and NS's replacements interspersed at various positions. Yes, this means NS is selectively overriding vanilla biomes at those positions in the biome table, but all the other positions... remain vanilla.

Terrablender works by defining regions in which mods like BOP or NS or WWEE have control over biome placement. For most mods this *will* include vanilla biomes, even if those biomes have been overhauled in some way. If you remove those, you're effectively handing Terrablender a gappy version of your own biome placement.

**Comment by nickcorrado** [comment]
Here's a way you can test my theory: make a fresh instance of Minecraft with WWEE and Terrablender installed. Set WWEE to Compatible mode and set the Terrablender config option I mentioned above to 0, then go in-game. I submit you'll find what I'm describing dialed up to 11: there will be no vanilla biomes *anywhere* in the world, and even the 'ocean' terrain will generate with land biome features. I tried it just now and that's exactly what I got.
